File: pkg/apis/training/tfjob_builder.go
package training

import (
	"fmt"
	"strings"

	"github.com/kubeflow/arena/pkg/apis/types"
	"github.com/kubeflow/arena/pkg/argsbuilder"
)

type TFJobBuilder struct {
	args      *types.SubmitTFJobArgs
	argValues map[string]interface{}
	argsbuilder.ArgsBuilder
}

func NewTFJobBuilder(args *types.SubmitTFJobArgs) *TFJobBuilder {
	if args == nil {
		args = &types.SubmitTFJobArgs{
			CleanPodPolicy:        "Running",
			CommonSubmitArgs:      DefaultCommonSubmitArgs,
			SubmitTensorboardArgs: DefaultSubmitTensorboardArgs,
		}
	}

	return &TFJobBuilder{
		args:        args,
		argValues:   map[string]interface{}{},
		ArgsBuilder: argsbuilder.NewSubmitTFJobArgsBuilder(args),
	}
}

func (b *TFJobBuilder) GetArgValues() map[string]interface{} {
	return b.argValues
}

func (b *TFJobBuilder) Name(name string) *TFJobBuilder {
	if name != "" {
		b.args.Name = name
	}
	return b
}

func (b *TFJobBuilder) Command(args []string) *TFJobBuilder {
	b.args.Command = strings.Join(args, " ")
	return b
}

func (b *TFJobBuilder) WorkingDir(dir string) *TFJobBuilder {
	if dir != "" {
		b.args.WorkingDir = dir
	}
	return b
}

func (b *TFJobBuilder) Envs(envs map[string]string) *TFJobBuilder {
	if envs != nil && len(envs) != 0 {
		envSlice := []string{}
		for key, value := range envs {
			envSlice = append(envSlice, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["env"] = &envSlice
	}
	return b
}

func (b *TFJobBuilder) GPUCount(count int) *TFJobBuilder {
	if count > 0 {
		b.args.GPUCount = count
	}
	return b
}

func (b *TFJobBuilder) Image(image string) *TFJobBuilder {
	if image != "" {
		b.args.Image = image
	}
	return b
}

func (b *TFJobBuilder) Tolerations(tolerations []string) *TFJobBuilder {
	b.argValues["toleration"] = &tolerations
	return b
}

func (b *TFJobBuilder) ConfigFiles(files map[string]string) *TFJobBuilder {
	if files != nil && len(files) != 0 {
		filesSlice := []string{}
		for localPath, containerPath := range files {
			filesSlice = append(filesSlice, fmt.Sprintf("%v:%v", localPath, containerPath))
		}
		b.argValues["config-file"] = &filesSlice
	}
	return b
}

func (b *TFJobBuilder) NodeSelectors(selectors map[string]string) *TFJobBuilder {
	if selectors != nil && len(selectors) != 0 {
		selectorsSlice := []string{}
		for key, value := range selectors {
			selectorsSlice = append(selectorsSlice, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["selector"] = &selectorsSlice
	}
	return b
}

func (b *TFJobBuilder) Annotations(annotations map[string]string) *TFJobBuilder {
	if annotations != nil && len(annotations) != 0 {
		s := []string{}
		for key, value := range annotations {
			s = append(s, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["annotation"] = &s
	}
	return b
}

func (b *TFJobBuilder) Labels(labels map[string]string) *TFJobBuilder {
	if labels != nil && len(labels) != 0 {
		s := []string{}
		for key, value := range labels {
			s = append(s, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["label"] = &s
	}
	return b
}

func (b *TFJobBuilder) EnableChief() *TFJobBuilder {
	b.args.UseChief = true
	return b
}

func (b *TFJobBuilder) ChiefCPU(cpu string) *TFJobBuilder {
	if cpu != "" {
		b.args.ChiefCpu = cpu
	}
	return b
}

func (b *TFJobBuilder) ChiefMemory(mem string) *TFJobBuilder {
	if mem != "" {
		b.args.ChiefMemory = mem
	}
	return b
}

func (b *TFJobBuilder) ChiefPort(port int) *TFJobBuilder {
	if port > 0 {
		b.args.ChiefPort = port
	}
	return b
}

func (b *TFJobBuilder) ChiefSelectors(selectors map[string]string) *TFJobBuilder {
	if selectors != nil && len(selectors) != 0 {
		s := []string{}
		for key, value := range selectors {
			s = append(s, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["chief-selector"] = &s
	}
	return b
}

func (b *TFJobBuilder) Datas(volumes map[string]string) *TFJobBuilder {
	if volumes != nil && len(volumes) != 0 {
		s := []string{}
		for key, value := range volumes {
			s = append(s, fmt.Sprintf("%v:%v", key, value))
		}
		b.argValues["data"] = &s
	}
	return b
}

func (b *TFJobBuilder) DataDirs(volumes map[string]string) *TFJobBuilder {
	if volumes != nil && len(volumes) != 0 {
		s := []string{}
		for key, value := range volumes {
			s = append(s, fmt.Sprintf("%v:%v", key, value))
		}
		b.argValues["data-dir"] = &s
	}
	return b
}

func (b *TFJobBuilder) EnableEvaluator() *TFJobBuilder {
	b.args.UseEvaluator = true
	return b
}

func (b *TFJobBuilder) EvaluatorCPU(cpu string) *TFJobBuilder {
	if cpu != "" {
		b.args.EvaluatorCpu = cpu
	}
	return b
}

func (b *TFJobBuilder) EvaluatorMemory(mem string) *TFJobBuilder {
	if mem != "" {
		b.args.EvaluatorMemory = mem
	}
	return b
}

func (b *TFJobBuilder) EvaluatorSelectors(selectors map[string]string) *TFJobBuilder {
	if selectors != nil && len(selectors) != 0 {
		s := []string{}
		for key, value := range selectors {
			s = append(s, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["evaluator-selector"] = &s
	}
	return b
}

func (b *TFJobBuilder) LogDir(dir string) *TFJobBuilder {
	if dir != "" {
		b.args.TrainingLogdir = dir
	}
	return b
}

func (b *TFJobBuilder) Priority(priority string) *TFJobBuilder {
	if priority != "" {
		b.args.PriorityClassName = priority
	}
	return b
}

func (b *TFJobBuilder) PsCount(count int) *TFJobBuilder {
	if count > 0 {
		b.args.PSCount = count
	}
	return b
}

func (b *TFJobBuilder) PsCPU(cpu string) *TFJobBuilder {
	if cpu != "" {
		b.args.PSCpu = cpu
	}
	return b
}

func (b *TFJobBuilder) PsGPU(gpu int) *TFJobBuilder {
	if gpu > 0 {
		b.args.PSGpu = gpu
	}
	return b
}

func (b *TFJobBuilder) PsImage(image string) *TFJobBuilder {
	if image != "" {
		b.args.PSImage = image
	}
	return b
}

func (b *TFJobBuilder) PsMemory(mem string) *TFJobBuilder {
	if mem != "" {
		b.args.PSMemory = mem
	}
	return b
}

func (b *TFJobBuilder) PsPort(port int) *TFJobBuilder {
	if port > 0 {
		b.args.PSPort = port
	}
	return b
}

func (b *TFJobBuilder) PsSelectors(selectors map[string]string) *TFJobBuilder {
	if selectors != nil && len(selectors) != 0 {
		s := []string{}
		for key, value := range selectors {
			s = append(s, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["ps-selector"] = &s
	}
	return b
}

func (b *TFJobBuilder) EnableRDMA() *TFJobBuilder {
	b.args.EnableRDMA = true
	return b
}

func (b *TFJobBuilder) SyncImage(image string) *TFJobBuilder {
	if image != "" {
		b.args.SyncImage = image
	}
	return b
}

func (b *TFJobBuilder) SyncMode(mode string) *TFJobBuilder {
	if mode != "" {
		b.args.SyncMode = mode
	}
	return b
}

func (b *TFJobBuilder) SyncSource(source string) *TFJobBuilder {
	if source != "" {
		b.args.SyncSource = source
	}
	return b
}

func (b *TFJobBuilder) EnableTensorboard() *TFJobBuilder {
	b.args.UseTensorboard = true
	return b
}

func (b *TFJobBuilder) TensorboardImage(image string) *TFJobBuilder {
	if image != "" {
		b.args.TensorboardImage = image
	}
	return b
}

func (b *TFJobBuilder) WorkerCPU(cpu string) *TFJobBuilder {
	if cpu != "" {
		b.args.WorkerCpu = cpu
	}
	return b
}

func (b *TFJobBuilder) WorkerImage(image string) *TFJobBuilder {
	if image != "" {
		b.args.WorkerImage = image
	}
	return b
}

func (b *TFJobBuilder) WorkerMemory(mem string) *TFJobBuilder {
	if mem != "" {
		b.args.WorkerMemory = mem
	}
	return b
}

func (b *TFJobBuilder) WorkerPort(port int) *TFJobBuilder {
	if port > 0 {
		b.args.WorkerPort = port
	}
	return b
}

func (b *TFJobBuilder) WorkerSelectors(selectors map[string]string) *TFJobBuilder {
	if selectors != nil && len(selectors) != 0 {
		s := []string{}
		for key, value := range selectors {
			s = append(s, fmt.Sprintf("%v=%v", key, value))
		}
		b.argValues["worker-selector"] = &s
	}
	return b
}

func (b *TFJobBuilder) WorkerCount(count int) *TFJobBuilder {
	if count > 0 {
		b.args.WorkerCount = count
	}
	return b
}
func (b *TFJobBuilder) ImagePullSecrets(secrets []string) *TFJobBuilder {
	if secrets != nil {
		b.argValues["image-pull-secret"] = &secrets
	}
	return b
}

func (b *TFJobBuilder) CleanPodPolicy(policy string) *TFJobBuilder {
	if policy != "" {
		b.args.CleanPodPolicy = policy
	}
	return b
}

func (b *TFJobBuilder) RoleSequence(roles []string) *TFJobBuilder {
	if roles != nil && len(roles) != 0 {
		b.argValues["role-sequence"] = strings.Join(roles, ",")
	}
	return b
}

func (b *TFJobBuilder) Build() (*Job, error) {
	for key, value := range b.argValues {
		b.AddArgValue(key, value)
	}
	if err := b.PreBuild(); err != nil {
		return nil, err
	}
	if err := b.ArgsBuilder.Build(); err != nil {
		return nil, err
	}
	return NewJob(b.args.Name, types.TFTrainingJob, b.args), nil
}

File: sdk/arena-java-sdk/src/main/java/com/github/kubeflow/arena/model/training/TFJobBuilder.java
package com.github.kubeflow.arena.model.training;

import com.github.kubeflow.arena.enums.TrainingJobType;
import com.github.kubeflow.arena.model.fields.BoolField;
import com.github.kubeflow.arena.model.fields.StringField;
import com.github.kubeflow.arena.model.fields.StringMapField;

import java.util.ArrayList;
import java.util.Map;

public class TFJobBuilder extends JobBuilder {

    public TFJobBuilder() {
        super(TrainingJobType.TFTrainingJob);
    }

    public TFJobBuilder workerCount(int count) {
        this.options.add(new StringField("--workers",String.valueOf(count)));
        return this;
    }

    public TFJobBuilder workerSelectors(Map<String, String> selectors) {
        this.options.add(new StringMapField("--worker-selector",selectors,"="));
        return this;
    }

    public TFJobBuilder workerPort(int port) {
        this.options.add(new StringField("--worker-port",String.valueOf(port)));
        return this;
    }

    public TFJobBuilder workerMemory(String memory) {
        this.options.add(new StringField("--worker-memory",memory));
        return this;
    }

    public TFJobBuilder workerMemoryLimit(String memory) {
        this.options.add(new StringField("--worker-memory-limit",memory));
        return this;
    }

    public TFJobBuilder workerImage(String image) {
        this.options.add(new StringField("--worker-image",image));
        return this;
    }

    public TFJobBuilder workerCPU(String cpu) {
        this.options.add(new StringField("--worker-cpu",cpu));
        return this;
    }

    public TFJobBuilder workerCPULimit(String cpu) {
        this.options.add(new StringField("--worker-cpu-limit",cpu));
        return this;
    }

    public TFJobBuilder psSelectors(Map<String, String> selectors) {
        this.options.add(new StringMapField("--ps-selector",selectors,"="));
        return this;
    }

    public TFJobBuilder psPort(int port) {
        this.options.add(new StringField("--ps-port",String.valueOf(port)));
        return this;
    }

    public TFJobBuilder psMemory(String memory) {
        this.options.add(new StringField("--ps-memory",memory));
        return this;
    }

    public TFJobBuilder psMemoryLimit(String memory) {
        this.options.add(new StringField("--ps-memory-limit",memory));
        return this;
    }

    public TFJobBuilder psImage(String image) {
        this.options.add(new StringField("--ps-image",image));
        return this;
    }

    public TFJobBuilder psCPU(String cpu) {
        this.options.add(new StringField("--ps-cpu",cpu));
        return this;
    }

    public TFJobBuilder psCPULimit(String cpu) {
        this.options.add(new StringField("--ps-cpu-limit",cpu));
        return this;
    }

    public TFJobBuilder psCount(int count) {
        this.options.add(new StringField("--ps",String.valueOf(count)));
        return this;
    }

    public TFJobBuilder evaluatorSelectors(Map<String, String> selectors) {
        this.options.add(new StringMapField("--evaluator-selector",selectors,"="));
        return this;
    }

    public TFJobBuilder evaluatorMemory(String memory) {
        this.options.add(new StringField("--evaluator-memory",memory));
        return  this;
    }

    public TFJobBuilder evaluatorMemoryLimit(String memory) {
        this.options.add(new StringField("--evaluator-memory-limit",memory));
        return  this;
    }

    public TFJobBuilder evaluatorCPU(String cpu) {
        this.options.add(new StringField("--evaluator-cpu",cpu));
        return this;
    }

    public TFJobBuilder evaluatorCPULimit(String cpu) {
        this.options.add(new StringField("--evaluator-cpu-limit",cpu));
        return this;
    }

    public TFJobBuilder enableEvaluator() {
        this.options.add(new BoolField("--evaluator"));
        return this;
    }

    public TFJobBuilder chiefSelectors(Map<String, String> selectors) {
        this.options.add(new StringMapField("--chief-selector",selectors,"="));
        return this;
    }

    public TFJobBuilder chiefPort(int port) {
        this.options.add(new StringField("--chief-port",String.valueOf(port)));
        return this;
    }

    public TFJobBuilder chiefMemory(String memory) {
        this.options.add(new StringField("--chief-memory",memory));
        return this;
    }

    public TFJobBuilder chiefMemoryLimit(String memory) {
        this.options.add(new StringField("--chief-memory-limit",memory));
        return this;
    }

    public TFJobBuilder chiefCPU(String cpu) {
        this.options.add(new StringField("--chief-cpu",cpu));
        return this;
    }

    public TFJobBuilder chiefCPULimit(String cpu) {
        this.options.add(new StringField("--chief-cpu-limit",cpu));
        return this;
    }

    public TFJobBuilder enableChief() {
        this.options.add(new BoolField("--chief"));
        return this;
    }

    /**
     * following functions invoke JobBuilder functions
     *
     *
     * **/


    public TFJobBuilder name(String name) {
        super.name(name);
        return this;
    }

    public TFJobBuilder image(String image) {
        super.image(image);
        return this;
    }

    public TFJobBuilder workers(int count) {
        super.workers(count);
        return this;
    }

    public TFJobBuilder imagePullSecrets(ArrayList<String> secrets) {
        super.imagePullSecrets(secrets);
        return this;
    }

    public TFJobBuilder gpus(int count) {
        super.gpus(count);
        return this;
    }

    public TFJobBuilder envs(Map<String, String> envs) {
        super.envs(envs);
        return this;
    }

    public TFJobBuilder nodeSelectors(Map<String, String> selectors) {
        super.nodeSelectors(selectors);
        return this;
    }

    public TFJobBuilder tolerations(ArrayList<String> tolerations) {
        super.tolerations(tolerations);
        return this;
    }

    public TFJobBuilder configFiles(Map<String, String> files) {
        super.configFiles(files);
        return this;
    }

    public TFJobBuilder annotations(Map<String, String> annotions) {
        super.annotations(annotions);
        return this;
    }

    public TFJobBuilder datas(Map<String, String> datas) {
        super.datas(datas);
        return this;
    }

    public TFJobBuilder dataDirs(Map<String, String> dataDirs) {
        super.dataDirs(dataDirs);
        return  this;
    }

    public TFJobBuilder logDir(String dir) {
        super.logDir(dir);
        return this;
    }

    public TFJobBuilder priority(String priority) {
        super.priority(priority);
        return  this;
    }

    public TFJobBuilder enableRDMA() {
        super.enableRDMA();
        return  this;
    }

    public TFJobBuilder syncImage(String image) {
        super.syncImage(image);
        return  this;
    }

    public TFJobBuilder syncMode(String mode) {
        super.syncMode(mode);
        return  this;
    }

    public TFJobBuilder syncSource(String source) {
        super.syncSource(source);
        return  this;
    }

    public TFJobBuilder enableTensorboard() {
        super.enableTensorboard();
        return this;
    }

    public TFJobBuilder tensorboardImage(String image) {
        super.tensorboardImage(image);
        return this;
    }

    public TFJobBuilder workingDir(String dir) {
        super.workingDir(dir);
        return this;
    }

    public TFJobBuilder retryCount(int count) {
        super.retryCount(count);
        return this;
    }

    public TFJobBuilder enableCoscheduling() {
        super.enableCoscheduling();
        return this;
    }

    public TFJobBuilder shell(String shell) {
        super.shell(shell);
        return this;
    }

    public TFJobBuilder command(String command) {
        this.command = command;
        return this;
    }

}
